cognite/client/_api/functions.py

```diff
@@ -521,44 +525,94 @@ def _create_session_and_return_nonce(
+def convert_file_path_to_module_path(root_path: str, file_path: Path) -> str:
+    """
+    Given a root directory and a full file path, this function will determine the relative
+    module path that can be used for importing.
+
+    Example:
+    If root_path is "/path/to/your/" and file_path is "/path/to/your/dir/module.py",
+    the function will return "dir.module".
+
+    Args:
+        root_path (str): The root directory from which the module import path should be determined.
+        file_path (Path): The full path to the Python file.
+    Returns:
+        str: The importable module path for the given file.
+    """
+    relative_path = Path(file_path).relative_to(root_path)  # get the relative path
+    module_path = ".".join(relative_path.with_suffix("").parts)
+    return module_path
+
+
+def _get_handle_function_node(handler_path: Path) -> ast.FunctionDef | None:
+    with handler_path.open() as f:
+        node = ast.parse(f.read())
+    for item in ast.walk(node):
+        if isinstance(item, ast.FunctionDef) and item.name == "handle":
+            return item
+    return None
+
+
+def _validate_function_from_ast(node: ast.FunctionDef) -> None:
+    # Ensure the function's name is "handle"
+    if node.name != "handle":
+        raise TypeError("Function referenced by function_handle must be named handle.")
 
+    # Validate function's arguments
+    arg_names = [arg.arg for arg in node.args.args]
+    allowed_args = {"data", "client", "secrets", "function_call_info"}
+    if not set(arg_names).issubset(allowed_args):
+        raise TypeError(
+            "Arguments to function referenced by function_handle must be a subset of (data, client, secrets, function_call_info)"
+        )
 
-def validate_function_folder(root_path: str, function_path: str) -> None:
+
+def validate_function_folder(root_path: str, function_path: str, check_imports: bool) -> None:
     if Path(function_path).suffix != ".py":
         raise TypeError(f"{function_path} is not a valid value for function_path. File extension must be .py.")
 
     function_path_full = Path(root_path, function_path)
     if not function_path_full.is_file():
         raise FileNotFoundError(f"No file found at location '{function_path}' in '{root_path}'.")
 
-    sys.path.insert(0, root_path)
+    handle_function_node = _get_handle_function_node(function_path_full)
+    if handle_function_node is None:
+        raise TypeError(f"{function_path} must contain a function named 'handle'.")
 
-    # Necessary to clear the cache if you have previously imported the module (this would have precedence over sys.path)
-    cached_handler_module = sys.modules.get("handler")
-    if cached_handler_module:
-        del sys.modules["handler"]
+    _validate_function_from_ast(handle_function_node)
 
-    module_path = convert_file_path_to_module_path(function_path)
-    handler = importlib.import_module(module_path)
+    # Opt-in import checks
+    if check_imports:
+        with _import_lock:
```

Comment:
@haakonvt 

Is this along the lines of what you envisioned? Not used thread locks before.

Reply:
Removed threadlock.